src/github.com/mozilla-services/pushgo/simplepush/worker.go

```diff
@@ -366,85 +366,76 @@ func (self *WorkerWS) Hello(header *RequestHeader, message []byte) (err error) {
-			}
-			reply := fmt.Sprintf(`{"messageType":%q,"uaid":%q,"status":429}`,
-				header.Type, uaid)
-			err = self.WriteText(reply)
-			self.stop()
-			return err
-		}
-		if !shouldRedirect {
-			goto registerDevice
-		}
-		if self.logger.ShouldLog(DEBUG) {
-			self.logger.Debug("worker", "Redirecting client", LogFields{
-				"rid": self.logID, "cmd": header.Type, "origin": origin})
-		}
-		reply := fmt.Sprintf(`{"messageType":%q,"uaid":%q,"status":307,"redirect":%q}`,
-			header.Type, uaid, origin)
-		err = self.WriteText(reply)
-		self.stop()
-		return err
+	if wroteReply {
+		w.stop()
+		return nil
 	}
-
-registerDevice:
-	// register any proprietary connection requirements
-	// alert the master of the new UAID.
-	// It's not a bad idea from a security POV to only send
-	// known args through to the server.
-	// blocking call back to the boss.
-	status, _ := ErrToStatus(
-		self.app.Server().Hello(self, []byte(request.PingData)))
-	if self.logger.ShouldLog(DEBUG) {
-		self.logger.Debug("worker", "sending response",
-			LogFields{"rid": self.logID, "cmd": "hello", "uaid": uaid})
-	}
-	reply := fmt.Sprintf(`{"messageType":%q,"uaid":%q,"status":%d}`,
-		header.Type, uaid, status)
-	if err = self.WriteText(reply); err != nil {
+	uaid := w.UAID()
+	if w.logger.ShouldLog(DEBUG) {
+		w.logger.Debug("worker", "sending response",
+			LogFields{"rid": w.logID, "cmd": "hello", "uaid": uaid})
+	}
+	reply := fmt.Sprintf(`{"messageType":%q,"uaid":%q,"status":200}`,
+		header.Type, uaid)
+	if err = w.WriteText(reply); err != nil {
 		if logWarning {
-			self.logger.Warn("dash", "Error writing client handshake", LogFields{
-				"rid": self.logID, "error": err.Error()})
+			w.logger.Warn("worker", "Error writing client handshake", LogFields{
+				"rid": w.logID, "error": err.Error()})
 		}
 		return err
 	}
-	self.metrics.Increment("updates.client.hello")
-	if self.logger.ShouldLog(INFO) {
-		self.logger.Info("dash", "Client successfully connected",
-			LogFields{"rid": self.logID})
+	w.metrics.Increment("updates.client.hello")
+	if w.logger.ShouldLog(INFO) {
+		w.logger.Info("worker", "Client successfully connected",
+			LogFields{"rid": w.logID})
 	}
-	self.state = WorkerActive
+	w.state = WorkerActive
 	// Get the lastAccessed time from wherever
-	return self.Flush(0, "", 0, "")
+	return w.Flush(0, "", 0, "")
+}
+
+// registerDevice adds the worker to the worker map and registers the
+// connecting client with the router.
+func (w *WorkerWS) registerDevice(header *RequestHeader,
+	request *HelloRequest) (wroteReply bool, err error) {
+
+	uaid, allowRedirect, err := w.handshake(request)
+	if err != nil {
+		return false, err
+	}
+	w.SetUAID(uaid)
+	if allowRedirect {
+		if wroteReply = w.checkRedirect(header); wroteReply {
+			return
+		}
+	}
+	// register any proprietary connection requirements
+	w.registerPropPing([]byte(request.PingData))
+	// Add the worker to the map and register with the router.
+	if added := w.app.AddWorker(uaid, w); added {
```

Comment:
I think `replaced` (or `!replaced`, in this case) conveys intent well, too..."if the worker isn't replacing an existing one, register with the router."